<a href="https://colab.research.google.com/github/andyotani/Machine-learning_26-27/blob/main/JS05/Tugas_Lab_2.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

**Tugas lab 2**

* Buatlah model klasfikasi Multinomial Naive Bayes dengan ketentuan,
    1.	Menggunakan data `spam.csv`
    2.	Fitur CountVectorizer dengan mengaktifkan stop_words
    3.	Evaluasi hasilnya
* Buatlah model klasfikasi Multinomial Naive Bayes dengan ketentuan,
    1.	Menggunakan data spam.csv
    2.	Fitur TF-IDF dengan mengaktifkan stop_words
    3.	Evaluasi hasilnya dan bandingkan dengan hasil pada Tugas no 2.
    4.	Berikan kesimpulan fitur mana yang terbaik pada kasus data spam.csv


In [44]:
# Persiapan Dataset
# Import Library
import pandas as pd
import numpy as np

from sklearn.model_selection import train_test_split

from sklearn.feature_extraction.text import CountVectorizer
from sklearn.feature_extraction.text import TfidfVectorizer

from sklearn.naive_bayes import MultinomialNB

from sklearn.metrics import accuracy_score
from sklearn.metrics import confusion_matrix
from sklearn.metrics import classification_report

In [45]:
# Load Dataset
data = pd.read_csv(
    "spam.csv",
    encoding="latin-1"
)

data.head()

,v1,v2,Unnamed: 2,Unnamed: 3,Unnamed: 4
0,ham,"Go until jurong point, crazy.. Available only ...",NaN,NaN,NaN
1,ham,Ok lar... Joking wif u oni...,NaN,NaN,NaN
2,spam,Free entry in 2 a wkly comp to win FA Cup fina...,NaN,NaN,NaN
3,ham,U dun say so early hor... U c already then say...,NaN,NaN,NaN
4,ham,"Nah I don't think he goes to usf, he lives aro...",NaN,NaN,NaN


Dataset memiliki beberapa kolom yang tidak digunakan, sehingga hanya digunakan:

* v1 → label spam/ham
* v2 → isi SMS

In [46]:
# Menghapus kolom tidak diperlukan
data = data[['v1','v2']]

data.columns = [
    'label',
    'SMS'
]

data.head()

,label,SMS
0,ham,"Go until jurong point, crazy.. Available only ..."
1,ham,Ok lar... Joking wif u oni...
2,spam,Free entry in 2 a wkly comp to win FA Cup fina...
3,ham,U dun say so early hor... U c already then say...
4,ham,"Nah I don't think he goes to usf, he lives aro..."


In [47]:
# Encoding Label

# Label diubah menjadi numerik:
# ham = 0
# spam = 1

data['label'] = data['label'].map(
    {
        'ham':0,
        'spam':1
    }
)

In [48]:
# Memisahkan fitur dan label
X = data['SMS']

y = data['label']

In [49]:
# Membagi Data Training dan Testing
# Data dibagi: 80% training dan 20% testing
X_train, X_test, y_train, y_test = train_test_split(
    X,
    y,
    test_size=0.2,
    random_state=50
)

### Percobaan 1 : Multinomial Naive Bayes + CountVectorizer

In [50]:
# Ekstraksi fitur CountVectorizer
# CountVectorizer mengubah teks SMS menjadi representasi jumlah kemunculan kata.
# Stop words diaktifkan untuk menghilangkan kata yang kurang memberikan informasi.
count_vectorizer = CountVectorizer(
    stop_words='english'
)


X_train_count = count_vectorizer.fit_transform(
    X_train
)


X_test_count = count_vectorizer.transform(
    X_test
)

In [51]:
# Membuat Model Multinomial Naive Bayes
model_count = MultinomialNB()


model_count.fit(
    X_train_count,
    y_train
)

MultinomialNB()

In [52]:
# Evaluasi Model
y_pred_count = model_count.predict(
    X_test_count
)


print(
    "Accuracy:",
    accuracy_score(
        y_test,
        y_pred_count
    )
)


print(
    confusion_matrix(
        y_test,
        y_pred_count
    )
)


print(
    classification_report(
        y_test,
        y_pred_count
    )
)

Accuracy: 0.9829596412556054
[[951   3]
 [ 16 145]]
              precision    recall  f1-score   support

           0       0.98      1.00      0.99       954
           1       0.98      0.90      0.94       161

    accuracy                           0.98      1115
   macro avg       0.98      0.95      0.96      1115
weighted avg       0.98      0.98      0.98      1115



Hasil CountVectorizer

Accuracy: 98.30%
Confusion Matrix: [[951   3]
 [ 16 145]]

 Interpretasi:

Terdapat 19 data yang mengalami kesalahan klasifikasi, yaitu 3 data ham diprediksi sebagai spam dan 16 data spam diprediksi sebagai ham.

### Percobaan 2 : Multinomial Naive Bayes + TF-IDF

In [53]:
# 1. Ekstraksi Fitur TF-IDF
# TF-IDF mempertimbangkan bobot pentingnya suatu kata berdasarkan frekuensi kemunculan kata tersebut.
tfidf_vectorizer = TfidfVectorizer(
    stop_words='english'
)


X_train_tfidf = tfidf_vectorizer.fit_transform(
    X_train
)


X_test_tfidf = tfidf_vectorizer.transform(
    X_test
)

In [54]:
# Membuat Model Naive Bayes
model_tfidf = MultinomialNB()


model_tfidf.fit(
    X_train_tfidf,
    y_train
)

MultinomialNB()

In [55]:
# Evaluasi Model
y_pred_tfidf = model_tfidf.predict(
    X_test_tfidf
)


print(
    "Accuracy:",
    accuracy_score(
        y_test,
        y_pred_tfidf
    )
)


print(
    confusion_matrix(
        y_test,
        y_pred_tfidf
    )
)


print(
    classification_report(
        y_test,
        y_pred_tfidf
    )
)

Accuracy: 0.9605381165919282
[[954   0]
 [ 44 117]]
              precision    recall  f1-score   support

           0       0.96      1.00      0.98       954
           1       1.00      0.73      0.84       161

    accuracy                           0.96      1115
   macro avg       0.98      0.86      0.91      1115
weighted avg       0.96      0.96      0.96      1115



Hasil TF-IDF

Accuracy: 96.05%
Confusion Matrix: [[954   0]
 [44 117]]

 Interpretasi:

Seluruh data ham berhasil diklasifikasikan dengan benar, tetapi masih terdapat 44 data spam yang salah diklasifikasikan sebagai ham.

### Perbandingan Hasil Model

| Metode Fitur | Algoritma | Accuracy |
|---|---|---|
| CountVectorizer + stop_words | Multinomial Naive Bayes | **98.30%** |
| TF-IDF + stop_words | Multinomial Naive Bayes | 96.05% |





### Kesimpulan
Berdasarkan hasil pengujian, model Multinomial Naive Bayes dengan fitur CountVectorizer dan stop_words menghasilkan performa terbaik dengan akurasi sebesar 98.30%, sedangkan penggunaan fitur TF-IDF menghasilkan akurasi sebesar 96.05%. Pada kasus klasifikasi SMS spam, CountVectorizer lebih optimal karena representasi berdasarkan frekuensi kemunculan kata mampu menangkap pola kata yang sering muncul pada pesan spam. Oleh karena itu, fitur terbaik untuk dataset spam.csv adalah CountVectorizer dengan stop_words menggunakan algoritma Multinomial Naive Bayes.

Dengan demikian, kombinasi terbaik untuk klasifikasi spam SMS pada dataset ini adalah CountVectorizer dengan stop_words menggunakan algoritma Multinomial Naive Bayes dengan akurasi sebesar 98.30%.